server/storage/mvcc/kvstore_test.go

```diff
@@ -178,6 +178,186 @@ func TestStorePut(t *testing.T) {
 	}
 }
 
+func TestStorePutDeleteWithLease(t *testing.T) {
+	lg := zaptest.NewLogger(t)
+	key := newTestKeyBytes(lg, revision{2, 0}, false)
+	kv := mvccpb.KeyValue{
+		Key:            []byte("foo"),
+		Value:          []byte("bar"),
+		CreateRevision: 2,
+		ModRevision:    2,
+		Version:        1,
+		Lease:          1,
+	}
+	kvb, err := kv.Marshal()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	tests := []struct {
+		// for put
+		r  indexGetResp
+		rr *rangeResp
+
+		// for delete
+		dr  indexRangeResp
+		drr *rangeResp
+	}{
+		{
+			indexGetResp{revision{}, revision{}, 0, ErrRevisionNotFound},
+			nil,
+
+			indexRangeResp{[][]byte{[]byte("foo")}, []revision{{2, 0}}},
+			&rangeResp{[][]byte{key}, [][]byte{kvb}},
+		},
+	}
+	for i, tt := range tests {
+		s := newFakeStore(lg)
+		s.le = lease.NewLessor(lg, s.b, nil, lease.LessorConfig{MinLeaseTTL: 5})
+		b := s.b.(*fakeBackend)
+		fi := s.kvindex.(*fakeIndex)
+
+		fi.indexGetRespc <- tt.r
+		if tt.rr != nil {
+			b.tx.rangeRespc <- *tt.rr
+		}
+
+		l, err := s.le.Grant(1, 10)
+		if err != nil {
+			t.Errorf("failed to grant lease with ttl(10s), err: %v", err)
+		}
+
+		s.Put([]byte("foo"), []byte("bar"), l.ID)
+
+		leaseAfterPut := s.le.Lookup(l.ID)
+		wantItemSet := []string{"foo"}
+		itemSet := leaseAfterPut.Keys()
+		if !reflect.DeepEqual(itemSet, wantItemSet) {
+			t.Errorf("#%d: leaseAfterPut.Keys() = %+v, want %+v", i, itemSet, wantItemSet)
+		}
+
+		itemMap := s.le.Items()
+		wantItemMap := []string{"foo"}
+		if !reflect.DeepEqual(itemMap, wantItemMap) {
+			t.Errorf("#%d: lessor.Items() = %+v, want %+v", i, itemMap, wantItemMap)
+		}
+
+		fi.indexRangeRespc <- tt.dr
+		if tt.rr != nil {
+			b.tx.rangeRespc <- *tt.drr
+		}
+
+		s.DeleteRange([]byte("foo"), nil)
+
+		leaseAfterDelete := s.le.Lookup(l.ID)
+		wantItemSet = []string{}
+		itemSet = leaseAfterDelete.Keys()
+
+		if !reflect.DeepEqual(itemSet, wantItemSet) {
+			t.Errorf("#%d: leaseAfterDelete.Keys() = %+v, want %+v", i, itemSet, wantItemSet)
+		}
+
+		itemMap = s.le.Items()
+		wantItemMap = []string{}
+		if !reflect.DeepEqual(itemMap, wantItemMap) {
+			t.Errorf("#%d: lessor.Items() = %+v, want %+v", i, itemMap, wantItemMap)
```

Comment:
maybe you can get away easier by just checking with `GetLease`? 

> assert.Equal(t, lease.NoLease, s.le.GetLease(lease.LeaseItem{Key: "foo"}))

That way we don't have to create another func on the lease api surface. Just to make everybody elses lives easier when we want to refactor it.

Reply:
updated